gprofiler/profilers/profiler_base.py

```diff
@@ -119,41 +157,162 @@ def _profiling_error_stack(
+
+    @property
+    def _is_profiling_spawning(self) -> bool:
+        return self._threads is not None
+
+    def _start_profiling_spawning(self, processes: List[Process]) -> None:
+        with self._submit_lock:
+            self._start_ts = time.monotonic()
+            # arbitrary high number of threads to make sure we can run profiling of many
+            # processes concurrently
+            self._threads = ThreadPoolExecutor(max_workers=999)
+            # TODO: add proc_events exit action to remove these
+            self._preexisting_pids = [p.pid for p in processes]
+
+    def _stop_profiling_spawning(self) -> None:
+        with self._submit_lock:
+            self._start_ts = None
+            assert self._threads is not None
+            threads = self._threads
+            # delete it before blocking on the exit of all threads (to ensure no new work
+            # is added)
+            self._threads = None
+            threads.shutdown()  # waits (although - all are done by now)
+            self._preexisting_pids = None
+
+    def _proc_exec_callback(self, tid: int, pid: int) -> None:
+        with contextlib.suppress(NoSuchProcess):
+            self._sched.enter(self._BACKOFF_INIT, 0, self._check_process, (Process(pid), self._BACKOFF_INIT))
+
+    def start(self) -> None:
+        super().start()
+
+        if self._profile_spawned_processes:
+            logger.debug(f"{self.__class__.__name__}: starting profiling spawning processes")
+            self._sched_thread.start()
+
+            try:
+                register_exec_callback(self._proc_exec_callback)
+            except Exception:
+                logger.warning("Failed to enable proc_events listener for executed processes", exc_info=True)
+            else:
+                self._enabled_proc_events_spawning = True
+
+    def stop(self) -> None:
+        super().stop()
 
+        if self._profile_spawned_processes:
+            if self._enabled_proc_events_spawning:
+                unregister_exec_callback(self._proc_exec_callback)
+                self._enabled_proc_events_spawning = False
+
+            self._sched_stop = True
+            self._sched_thread.join()
+
+    def snapshot(self) -> ProcessToProfileData:
+        results = super().snapshot()
+
+        # wait for one duration, in case snapshot() found no processes
+        self._stop_event.wait(self._duration)
+
+        self._stop_profiling_spawning()
+        results_spawned = self._wait_for_profiles(self._futures)
+        self._futures = {}
+
+        # should not intersect
+        assert set(results).intersection(results_spawned) == set()
+        results.update(results_spawned)
         return results
+
+    def _sched_thread_run(self) -> None:
+        while not (self._stop_event.is_set() or self._sched_stop):
+            self._sched.run()
+            self._stop_event.wait(0.1)
+
+    def _check_process(self, process: Process, interval: float) -> None:
```

Comment:
Clear `self._sched` if we're stopping, so that `self._sched.run()` exits faster.
```suggestion
    def _clear_shced(self):
        for event in self._sched.queue:
            self._sched.cancel(event)

    def _check_process(self, process: Process, interval: float) -> None:
        with contextlib.suppress(NoSuchProcess):
            if not self._is_profiling_spawning:
                self._clear_sched()
                return

            if not is_process_running(process) or process.ppid() == os.getpid():
```

Reply:
The `clear_sched` call needn't be in suppress

Reply:
I agree

Reply:
```suggestion
    def _clear_shced(self):
        for event in self._sched.queue:
            self._sched.cancel(event)
            
    def _check_process(self, process: Process, interval: float) -> None:
        if not self._is_profiling_spawning:
            self._clear_sched()
            return
        
        with contextlib.suppress(NoSuchProcess):
            if not is_process_running(process) or process.ppid() == os.getpid():```

Reply:
Premature optimization if I've ever seen one